File: LeetCode/LeetCode_Easy/TypeScript/LinkedListCycle.ts
// Definition for singly-linked list.
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function hasCycle(head: ListNode | null): boolean {
  if (head === null || head.next === null) {
    return false;
  }

  let rabbit: ListNode | null | undefined = head.next;
  let turtle: ListNode | null | undefined = head;

  while (rabbit !== turtle) {
    if (rabbit === null || rabbit.next === null) {
      return false;
    }

    turtle = turtle?.next;
    rabbit = rabbit.next.next;
  }
  return true;
}

File: LeetCode/LeetCode_Easy/C++/LinkedListCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == NULL || head->next == NULL) {
            return false;
        }

        ListNode *rabbit = head->next;
        ListNode *turtle = head;

        while (rabbit != NULL && rabbit->next != NULL) {
            if (rabbit->next->next == NULL || turtle->next == NULL) {
                return false;
            }

            if (rabbit == turtle) {
                return true;
            }
            
            rabbit = rabbit->next->next;
            turtle = turtle->next;
        }

        return rabbit == turtle;
    }
};